devito/types/equation.py

```diff
@@ -140,10 +142,26 @@ def _flatten(self):
         else:
             return [self]
 
-    @property
+    @cached_property
     def subdomain(self):
         """The SubDomain in which the Eq is defined."""
-        return self._subdomain
+        if self._subdomain is not None:
+            return self._subdomain
+
+        from devito.symbolics.search import retrieve_functions  # noqa
+
+        funcs = retrieve_functions(self)
+        subdomains = {f.grid for f in funcs} - {None}
+        subdomains = {g for g in subdomains if g.is_SubDomain}
```

Comment:
You're lacking a test -- if the Eq had a Function not carrying a Grid, you'd get an AttributeError here.

I wonder whether it might be convenient to create a utility function along the lines of `retrieve_cartesian_discretization(functions, cls=Grid)` which returns the _unique_ cartesian discretization of type `cls` among `functions`, or throws an exception otherwise. Could such a function be used, besides here, in operator.py as well? and maybe in interpolators.py too in place of https://github.com/devitocodes/devito/blob/ec7b22aea547f4f5a34455a7649672dc093e0935/devito/operations/interpolators.py#L36 ?

Reply:
> if the Eq had a Function not carrying a Grid, you'd get an AttributeError here.
This isn't correct. In that case `f.grid` would be `None`, hence the `- {None}`.

I think the utility function isn't currently needed? I can only think of two concrete uses of this kind of search, and they are both substantially different

Reply:
ah ok alright